storage/innobase/lock/lock0lock.cc

```diff
@@ -5364,6 +5374,23 @@ static void lock_cancel_waiting_and_release(lock_t *lock)
   lock_wait_end(trx);
   trx->mutex_unlock();
 }
+#ifdef WITH_WSREP
+/* if trx is waiting for a lock, cancel the waiting
+*/
+void lock_sys_t::cancel_lock_wait_for_trx(trx_t *trx)
+{
+  lock_sys.wr_lock(SRW_LOCK_CALL);
+  mysql_mutex_lock(&lock_sys.wait_mutex);
+  if (lock_t *lock= trx->lock.wait_lock)
+  {
+    /* check if victim is still waiting */
+    if (lock->is_waiting())
+      lock_cancel_waiting_and_release(lock);
+  }
+  mysql_mutex_unlock(&lock_sys.wait_mutex);
+  lock_sys.wr_unlock();
```

Comment:
For scalability, it is better to release the more contended resource first. `lock_wait_wsrep_kill()` is invoking `lock_sys.wr_unlock()` before the `mysql_mutex_unlock(&lock_sys.wait_mutex)` for a reason.